internal/backend/oauth/webhook.go

```diff
@@ -1,134 +1,167 @@
+	intg := r.PathValue("intg")
+
+	l := s.logger.With(zap.String("urlPath", r.URL.Path), zap.String("intg", intg))
 
 	// Sanity check: the specified OAuth consumer is registered.
 	o := s.svcs.OAuth()
-	ctx := context.Background()
-	_, _, err := o.Get(ctx, id)
-	if err != nil {
-		s.logger.Warn("OAuth config retrieval error",
-			zap.String("urlPath", r.URL.Path),
-			zap.Error(err),
-		)
+	ctx := r.Context()
+	if _, _, err := o.Get(ctx, r.PathValue("intg")); err != nil {
+		l.Warn("OAuth config retrieval error", zap.Error(err))
 		http.Error(w, "Bad request: unrecognized config ID", http.StatusBadRequest)
 		return
 	}
 
-	switch action {
-	// Start a 3-legged OAuth v2 user flow.
-	case "start":
-		s.startFlow(ctx, w, r, o, id)
-	// Accept a temporary authorization code via redirection,
-	// and convert it into a refresh token / user access token.
-	case "redirect":
-		s.exchangeCode(ctx, w, r, o, id)
-	// Error: bad request with an unrecognized action.
-	default:
-		s.logger.Warn("Unrecognized OAuth webhook action",
-			zap.String("urlPath", r.URL.Path),
-		)
-		http.Error(w, "Bad request: unrecognized webhook action", http.StatusBadRequest)
+	cid, err := sdktypes.ParseConnectionID(r.URL.Query().Get("cid"))
+	if err != nil {
+		l.Warn("Failed to parse connection ID", zap.Error(err))
+		http.Error(w, "Bad request: invalid connection ID", http.StatusBadRequest)
+		return
 	}
-}
 
-func (s *svc) startFlow(ctx context.Context, w http.ResponseWriter, r *http.Request, o sdkservices.OAuth, id string) {
-	startURL, err := o.StartFlow(ctx, id)
+	startURL, err := o.StartFlow(ctx, intg, cid)
 	if err != nil {
-		s.logger.Error("OAuth start flow error",
-			zap.String("urlPath", r.URL.Path),
-			zap.Error(err),
-		)
+		l.Error("OAuth start flow error", zap.Error(err))
 		http.Error(w, "Bad request: start flow error", http.StatusBadRequest)
 		return
 	}
 	http.Redirect(w, r, startURL, http.StatusFound)
 }
 
-func (s *svc) exchangeCode(ctx context.Context, w http.ResponseWriter, r *http.Request, o sdkservices.OAuth, id string) {
-	destURL := fmt.Sprintf("/%s/oauth?%s", id, r.URL.RawQuery)
-	destURL = regexp.MustCompile(`&?code=[^&]*`).ReplaceAllString(destURL, "")
-	destURL = regexp.MustCompile(`&?state=[^&]*`).ReplaceAllString(destURL, "")
+func (s *svc) exchange(w http.ResponseWriter, r *http.Request) {
+	intg := r.PathValue("intg")
+	intgSymbol, err := sdktypes.ParseSymbol(intg)
+	if err != nil {
+		http.Error(w, "invalid integration name", http.StatusBadRequest)
+		return
+	}
 
-	// Read our state value from the the redirected request.
-	state := r.FormValue("state")
-	if state == "" {
-		s.logger.Warn("OAuth redirect request without state parameter",
-			zap.String("urlPath", r.URL.Path),
-			zap.Any("form", r.Form),
-		)
-		destURL += "&error=" + url.QueryEscape("Missing state parameter")
-		http.Redirect(w, r, destURL, http.StatusFound)
+	l := s.logger.With(zap.String("url_path", r.URL.Path), zap.String("intg", intg), zap.Any("form", r.Form))
```

Comment:
Printing the form in the log is a bad idea. Either it's empty - so it's not useful, or it contains sensitive oauth data - which shouldn't be leaked into our logs

Reply:
removed